File: 모범 코드/7일차/7-2. 불..java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class main {
    static int arr[][];
    static int dist[][];
    static int R, C, sx, sy;
    static int dx[] = { -1, 0, 1, 0 };
    static int dy[] = { 0, 1, 0, -1 };

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();
		
        int T = Integer.parseInt(br.readLine());
        while(T-- > 0) {
            st = new StringTokenizer(br.readLine());
            C = Integer.parseInt(st.nextToken());
            R = Integer.parseInt(st.nextToken());
			
            arr = new int[R + 2][C + 2];
            dist = new int[R + 2][C + 2];
			
            Queue<Integer> que = new LinkedList<>();

            for(int i = 1; i <= R; i++) {
                String str = br.readLine();
                for(int j = 1; j <= C; j++) {
                    char tmp = str.charAt(j - 1);
					if (tmp == '@'){
						sx = i;
						sy = j;
					}
					if(tmp == '*'){
						que.offer(i);
						que.offer(j);
						que.offer(1);
					}
                    if(tmp == '.' || tmp == '@' || tmp == '*') arr[i][j] = 1;
                }
            }

			que.offer(sx);
			que.offer(sy);
			que.offer(-1);
            while(!que.isEmpty()) {
				int x = que.poll();
				int y = que.poll();
				int d = que.poll();
				
				if(arr[x][y] == 0 || dist[x][y] != 0) continue;
				dist[x][y] = d;
				
                for (int i = 0; i < 4; i++) {
					int d_ = d > 0 ? d+1 : d-1;
					que.offer(x + dx[i]);
					que.offer(y + dy[i]);
					que.offer(d_);
                }
            }
			
			int ans = - R * C - 1;
			for(int i=1; i<=R; i++){
				if(dist[i][1] < 0) ans = Math.max(ans, dist[i][1]);
				if(dist[i][C] < 0) ans = Math.max(ans, dist[i][C]);
			}
			for(int j=1; j<=C; j++){
				if(dist[1][j] < 0) ans = Math.max(ans, dist[1][j]);
				if(dist[R][j] < 0) ans = Math.max(ans, dist[R][j]);
			}
            if (ans == -R*C-1)
                sb.append("IMPOSSIBLE\n");
            else
                sb.append((-ans) + "\n");
        }
        System.out.println(sb.toString());
    }

}

File: 모범 코드/7일차/7-2. 불.py
from collections import deque

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

for _ in range(int(input())):
	W, H = map(int, input().split())
	dist = [[0] * W for _ in range(H)]
	arr = []
	que = deque()
	ans = W * H + 1
	for h in range(H):
		arr.append(list(input()))
		for w in range(W):
			if arr[h][w] == '@':
				arr[h][w] = '.'
				x = h
				y = w
			if arr[h][w] == '*':
				que.append((h, w, 1))
				arr[h][w] = '.'
	que.append((x, y, -1))
	
	while que:
		x, y, d = que.popleft()
		d_ = d+1 if d > 0 else d-1
		
		if x < 0 or x >= H or y < 0 or y >= W or arr[x][y] != '.' or dist[x][y] != 0: continue
		if d < 0 and (x == 0 or x == H-1 or y == 0 or y == W-1):
			ans = min(ans, -d)
		
		dist[x][y] = d
		
		for i in range(4):
			que.append((x+dx[i], y+dy[i], d_))
			
	if ans == H*W+1: print("IMPOSSIBLE")
	else: print(ans)
